eureka2-client/src/main/java/com/netflix/eureka2/client/resolver/RoundRobinServerResolver.java

```diff
@@ -0,0 +1,143 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Notification;
+import rx.Observable;
+import rx.functions.Func1;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A Server Resolver that uses a round-robin strategy to select between servers. The {@link #resolve()} will
+ * not emit any elements until the resolver have warmed up for the first time, i.e. the resolver has been loaded
+ * with an initial list of servers. The warm up time can be changed from defaults by calling
+ * {@link #withWarmUpConfiguration(int, TimeUnit)}.
+ *
+ * @author David Liu
+ */
+public class RoundRobinServerResolver implements ServerResolver {
+
+    private static final Logger logger = LoggerFactory.getLogger(RoundRobinServerResolver.class);
+
+    private static final Exception SERVER_CACHE_EMPTY_EXCEPTION = new NoSuchElementException("No servers available for this resolver");
+
+    private final int warmUpTimeout;
+    private final TimeUnit timeUnit;
+
+    private final Observable<ChangeNotification<Server>> serverSource;
+    private final AtomicReference<List<Server>> serverCacheRef;
+    private final AtomicInteger positionRef;
+
+    protected RoundRobinServerResolver(Server... servers) {
+        this(sourceFromList(servers), 10, TimeUnit.SECONDS);
+    }
+
+    protected RoundRobinServerResolver(Observable<ChangeNotification<Server>> serverSource) {
+        this(serverSource, 10, TimeUnit.SECONDS);
+    }
+
+    protected RoundRobinServerResolver(Observable<ChangeNotification<Server>> serverSource, int warmUpTimeout, TimeUnit timeUnit) {
+        this.serverSource = serverSource;
+        this.warmUpTimeout = warmUpTimeout;
+        this.timeUnit = timeUnit;
+        this.serverCacheRef = new AtomicReference<List<Server>>(new ArrayList<Server>());
+        this.positionRef = new AtomicInteger(new Random().nextInt(100));
+    }
+
+    public RoundRobinServerResolver withWarmUpConfiguration(int newWarmUpTimeout, TimeUnit newTimeUnit) {
+        return new RoundRobinServerResolver(serverSource, newWarmUpTimeout, newTimeUnit);
+    }
+
+    @Override
+    public void close() {
+        serverCacheRef.get().clear();
+    }
+
+    @Override
+    public Observable<Server> resolve() {
+        return refreshServerCache().concatMap(new Func1<List<Server>, Observable<? extends Server>>() {
+            @Override
+            public Observable<? extends Server> call(List<Server> servers) {
+                if (servers.isEmpty()) {
+                    return Observable.error(SERVER_CACHE_EMPTY_EXCEPTION);
+                }
+
+                int currentPos = positionRef.getAndIncrement();
```

Comment:
I would do this module servers.size() to avoid Integer overflow. I know it is very unlikely, but still it is better to be strict if there is no cost to it.